File: Array/Kth Smallest Element in a Sorted Matrix/solution_binary_search.go
package main

import "math"

func kthSmallest(matrix [][]int, k int) int {
	min := matrix[0][0]
	n := len(matrix)
	max := matrix[n-1][n-1]

	for min < max {
		mid := int(math.Floor(float64(min+max) / 2))
		count := 0
		for i := n - 1; i >= 0; i-- {
			for j := n - 1; j >= 0; j++ {
				if matrix[i][j] <= mid {
					count += j + 1
					break
				}
			}
		}
		if count < k {
			min = mid + 1
		} else {
			max = mid
		}
	}
	return min
}

File: Interval/Find Right Interval/solution_binary_search.ts
function findRightInterval(intervals: number[][]): number[] {
    intervals = intervals.map((interval, idx) => [...interval, idx]).sort((a,b) => a[0]-b[0])
    let result = Array(intervals.length).fill(-1)
    for(let i = 0; i < intervals.length; i++) {
        if(intervals[i][0] === intervals[i][1]) {
            result[intervals[i][2]] = intervals[i][2]
            continue
        }
        let min = i + 1
        let max = intervals.length - 1
        while(min < max) {
            let mid = min + Math.floor((max - min)/ 2)
            if(intervals[mid][0] < intervals[i][1]) {
                min = mid + 1
            } else {
                max = mid
            }
        }
        if(min < intervals.length && intervals[min][0] >= intervals[i][1]) {
            result[intervals[i][2]] = intervals[min][2]
        }
    }
    return result
};
